rust/lance-index/src/vector/pq.rs

```diff
@@ -333,89 +333,64 @@ where
-                        dim,
-                        num_bits,
-                        num_sub_vectors,
-                        idx,
-                    )
-                })
-                .collect::<Vec<_>>();
-            let flatten_data =
-                fsl.values()
-                    .as_any()
-                    .downcast_ref::<T::ArrayType>()
-                    .ok_or(Error::Index {
-                        message: format!(
-                            "Expect to be a float vector array, got: {:?}",
-                            fsl.value_type()
-                        ),
-                        location: location!(),
-                    })?;
-
-            let flatten_values = flatten_data.as_slice();
-            let capacity = num_sub_vectors * num_rows;
-            let mut builder: Vec<u8> = vec![0; capacity];
-            // Dimension of each sub-vector.
-            let sub_dim = dim / num_sub_vectors;
-            for i in 0..num_rows {
-                let row_offset = i * dim;
-
-                for sub_idx in 0..num_sub_vectors {
-                    let offset = row_offset + sub_idx * sub_dim;
-                    let sub_vector = &flatten_values[offset..offset + sub_dim];
-                    let centroids = all_centroids[sub_idx];
-
-                    let dist_iter = match metric_type {
-                        MetricType::L2 | MetricType::Cosine => {
-                            l2_distance_batch(sub_vector, centroids, sub_dim)
-                        }
-                        MetricType::Dot => dot_distance_batch(sub_vector, centroids, sub_dim),
-                        _ => panic!(
-                            "ProductQuantization: metric type {} not supported",
-                            metric_type
-                        ),
-                    };
-                    let code = argmin(dist_iter).ok_or(Error::Index {
-                        message: format!(
-                            "Failed to assign PQ code: {}, sub-vector={:#?}",
-                            "it is likely that distance is NaN or Inf", sub_vector
-                        ),
-                        location: location!(),
-                    })? as u8;
-                    builder[i * num_sub_vectors + sub_idx] = code as u8;
-                }
-            }
-            Ok::<UInt8Array, Error>(UInt8Array::from(builder))
-        })
-        .await??;
+        let distance_type = self.metric_type;
+
+        let flatten_data =
+            fsl.values()
+                .as_any()
+                .downcast_ref::<T::ArrayType>()
+                .ok_or(Error::Index {
+                    message: format!(
+                        "Expect to be a float vector array, got: {:?}",
+                        fsl.value_type()
+                    ),
+                    location: location!(),
+                })?;
+
+        let values = flatten_data
+            .as_slice()
+            .par_chunks(dim)
```

Comment:
nit: would it be more effcient to do `.chunk_exact(dim). into_par_iter().map(...)`?

Reply:
It does not compile in that way

```
error[E0599]: the method `into_par_iter` exists for struct `ChunksExact<'_, <T as ArrowFloatType>::Native>`, but its trait bounds were not satisfied
    --> rust/lance-index/src/vector/pq.rs:356:14
     |
353  |           let values = flatten_data
     |  ______________________-
354  | |             .as_slice()
355  | |             .chunks_exact(dim)
356  | |             .into_par_iter()
     | |_____________-^^^^^^^^^^^^^
     |
    ::: /Users/lei/.rustup/toolchains/stable-aarch64-apple-darwin/lib/rustlib/src/rust/library/core/src/slice/iter.rs:1828:1
     |
1828 |   pub struct ChunksExact<'a, T: 'a> {
     |   --------------------------------- doesn't satisfy `_: IntoParallelIterator` or `_: ParallelIterator`
     |
     = note: the following trait bounds were not satisfied:
             `std::slice::ChunksExact<'_, <T as lance_arrow::ArrowFloatType>::Native>: rayon::iter::ParallelIterator`
             which is required by `std::slice::ChunksExact<'_, <T as lance_arrow::ArrowFloatType>::Native>: rayon::iter::IntoParallelIterator`
             `&std::slice::ChunksExact<'_, <T as lance_arrow::ArrowFloatType>::Native>: rayon::iter::ParallelIterator`
             which is required by `&std::slice::ChunksExact<'_, <T as lance_arrow::ArrowFloatType>::Native>: rayon::iter::IntoParallelIterator`
             `&mut std::slice::ChunksExact<'_, <T as lance_arrow::ArrowFloatType>::Native>: rayon::iter::ParallelIterator`
             which is required by `&mut std::slice::ChunksExact<'_, <T as lance_arrow::ArrowFloatType>::Native>: rayon::iter::IntoParallelIterator`

warning: unused import: `rayon::prelude`
  --> rust/lance-index/src/vector/pq.rs:18:5
```